File: src/nssh/cli/common/batch.py
"""Shared batch operation utilities for nssh CLI commands."""

from __future__ import annotations

import csv
import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Callable, List, Optional, Type, TypeVar

from nssh.core.ui.console import get_console

console = get_console()

T = TypeVar("T")

# Security: Maximum batch file size to prevent memory exhaustion
MAX_BATCH_FILE_SIZE = 10 * 1024 * 1024  # 10 MB


@dataclass
class BatchResult:
    """Result of a batch operation."""

    added: int = 0
    skipped: int = 0
    failed: int = 0
    errors: List[str] = field(default_factory=list)

    @property
    def total_processed(self) -> int:
        return self.added + self.skipped + self.failed

    @property
    def removed(self) -> int:
        """Alias for added - used in removal operations for semantic clarity."""
        return self.added

    def has_failures(self) -> bool:
        return self.failed > 0


@dataclass
class HostEntry:
    """Entry for batch host operations (add/edit).

    The `host` field is optional. If omitted, the SSH alias is derived
    from the hostname by taking the first segment before the first dot.
    For example, 'rpi-b.home.arpa' becomes 'rpi-b'.
    """

    hostname: str
    user: Optional[str] = None
    port: Optional[int] = None
    context: Optional[str] = None
    password: Optional[str] = None
    host: Optional[str] = None

    @property
    def alias(self) -> str:
        """Return the SSH Host alias for this entry.

        If `host` is explicitly set, use it. Otherwise derive from hostname
        by splitting on the first dot.
        """
        if self.host:
            return self.host
        return self.hostname.split(".")[0]

    @classmethod
    def from_dict(cls, data: dict) -> "HostEntry":
        """Create HostEntry from a dictionary."""
        return cls(
            hostname=data.get("hostname", ""),
            user=data.get("user") or None,
            port=int(data["port"]) if data.get("port") else None,
            context=data.get("context") or None,
            password=data.get("password") or None,
            host=data.get("host") or None,
        )


@dataclass
class EditEntry:
    """Entry for batch edit operations."""

    hostname: str
    user: Optional[str] = None  # None = no change
    port: Optional[int] = None  # None = no change
    password: Optional[str] = None  # None = no change

    @classmethod
    def from_dict(cls, data: dict) -> "EditEntry":
        """Create EditEntry from a dictionary."""
        return cls(
            hostname=data.get("hostname", ""),
            user=data.get("user") or None,
            port=int(data["port"]) if data.get("port") else None,
            password=data.get("password") or None,
        )


def is_batch_file(arg: str) -> bool:
    """Detect if argument is a batch file based on extension.

    Args:
        arg: Command line argument (hostname or file path)

    Returns:
        True if arg ends with .txt, .csv, or .json (case-insensitive)
    """
    return arg.lower().endswith((".txt", ".csv", ".json"))


def parse_txt_file(path: Path) -> List[str]:
    """Parse a .txt file with one hostname per line.

    Args:
        path: Path to .txt file

    Returns:
        List of hostnames (empty lines and comments ignored)
    """
    hostnames = []
    with open(path, "r") as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith("#"):
                hostnames.append(line)
    return hostnames


def parse_csv_file(path: Path, entry_cls: Type[T]) -> List[T]:
    """Parse a .csv file with headers into entry objects.

    Args:
        path: Path to .csv file
        entry_cls: Dataclass with from_dict() classmethod

    Returns:
        List of entry objects
    """
    entries = []
    with open(path, "r", newline="") as f:
        reader = csv.DictReader(f)
        for row in reader:
            entries.append(entry_cls.from_dict(row))  # type: ignore[attr-defined]
    return entries


def parse_json_file(path: Path, entry_cls: Type[T]) -> List[T]:
    """Parse a .json file (array of objects) into entry objects.

    Args:
        path: Path to .json file
        entry_cls: Dataclass with from_dict() classmethod

    Returns:
        List of entry objects
    """
    with open(path, "r") as f:
        data = json.load(f)

    if not isinstance(data, list):
        raise ValueError("JSON file must contain an array of objects")

    return [entry_cls.from_dict(item) for item in data]  # type: ignore[attr-defined]


def parse_batch_file(
    path: str, entry_cls: Type[T], txt_to_entry: Optional[Callable[[str], T]] = None
) -> List[T]:
    """Parse a batch file into entry objects based on extension.

    Args:
        path: Path to batch file (.txt, .csv, or .json)
        entry_cls: Dataclass with from_dict() classmethod
        txt_to_entry: Optional converter for .txt lines (if None, txt parsing fails)

    Returns:
        List of entry objects

    Raises:
        ValueError: If file format is unsupported or invalid
        FileNotFoundError: If file doesn't exist
    """
    file_path = Path(path)

    if not file_path.exists():
        raise FileNotFoundError(f"Batch file not found: {path}")

    # Security: Limit file size to prevent memory exhaustion
    file_size = file_path.stat().st_size
    if file_size > MAX_BATCH_FILE_SIZE:
        raise ValueError(
            f"Batch file too large ({file_size / 1024 / 1024:.1f} MB). "
            f"Maximum allowed: {MAX_BATCH_FILE_SIZE / 1024 / 1024:.0f} MB"
        )

    ext = file_path.suffix.lower()

    if ext == ".txt":
        if txt_to_entry is None:
            raise ValueError(".txt format requires a txt_to_entry converter")
        hostnames = parse_txt_file(file_path)
        return [txt_to_entry(h) for h in hostnames]
    elif ext == ".csv":
        return parse_csv_file(file_path, entry_cls)
    elif ext == ".json":
        return parse_json_file(file_path, entry_cls)
    else:
        raise ValueError(f"Unsupported file format: {ext}")


def display_batch_errors(result: BatchResult) -> None:
    """Display errors from a batch operation.

    Args:
        result: BatchResult with errors
    """
    if result.errors:
        for error in result.errors[:10]:
            console.print(f"  [red]{error}[/red]")
        if len(result.errors) > 10:
            console.print(f"  [dim]... and {len(result.errors) - 10} more[/dim]")


def validate_hostname(hostname: str) -> Optional[str]:
    """Validate a hostname/FQDN format.

    Args:
        hostname: Hostname to validate

    Returns:
        Error message if invalid, None if valid
    """
    if not hostname:
        return "hostname is required"
    if not hostname.strip():
        return "hostname cannot be empty"
    # Basic validation - hostname shouldn't have spaces or invalid chars
    if " " in hostname:
        return f"invalid hostname format: {hostname}"
    return None

File: internal/cli/host/batch.go
package host

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// MaxBatchFileSize is the maximum size of a batch file (10MB).
// This prevents memory exhaustion from maliciously large files.
const MaxBatchFileSize = 10 * 1024 * 1024

// BatchResult tracks the outcome of a batch operation.
type BatchResult struct {
	Added   int
	Skipped int
	Failed  int
	Errors  []string
}

// TotalProcessed returns the total number of entries processed.
func (r *BatchResult) TotalProcessed() int {
	return r.Added + r.Skipped + r.Failed
}

// HasFailures returns true if any entries failed.
func (r *BatchResult) HasFailures() bool {
	return r.Failed > 0
}

// HostEntry represents a host from a batch file.
type HostEntry struct {
	Host     string `json:"host"`               // The SSH Host identifier (unique within a context)
	HostName string `json:"hostname,omitempty"` // The SSH HostName (resolved address, FQDN or IP)
	User     string `json:"user,omitempty"`
	Port     int    `json:"port,omitempty"`
	Context  string `json:"context,omitempty"`
	Password string `json:"password,omitempty"`
}

// IsBatchFile detects if an argument is a batch file based on extension.
// Supported formats: CSV (.csv) and JSON (.json).
func IsBatchFile(arg string) bool {
	ext := strings.ToLower(filepath.Ext(arg))
	switch ext {
	case ".csv", ".json":
		return true
	default:
		return false
	}
}

// ValidateHostname validates a hostname format.
// Returns an error message if invalid, empty string if valid.
func ValidateHostname(hostname string) string {
	if hostname == "" {
		return "hostname is required"
	}
	hostname = strings.TrimSpace(hostname)
	if hostname == "" {
		return "hostname cannot be empty"
	}
	if strings.ContainsAny(hostname, " \t\n\r") {
		return fmt.Sprintf("invalid hostname format: %s", hostname)
	}
	return ""
}

// ParseBatchFile parses a batch file into HostEntry objects based on extension.
// Supported formats: CSV (.csv) and JSON (.json).
func ParseBatchFile(path string) ([]HostEntry, error) {
	// Check file exists
	fi, err := os.Stat(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("batch file not found: %s", path)
		}
		return nil, fmt.Errorf("stat batch file: %w", err)
	}

	// Check file size
	if fi.Size() > MaxBatchFileSize {
		return nil, fmt.Errorf("batch file too large (%.1f MB). Maximum allowed: %.0f MB",
			float64(fi.Size())/(1024*1024),
			float64(MaxBatchFileSize)/(1024*1024))
	}

	ext := strings.ToLower(filepath.Ext(path))
	switch ext {
	case ".csv":
		return parseCsvFile(path)
	case ".json":
		return parseJsonFile(path)
	default:
		return nil, fmt.Errorf("unsupported file format: %s (use .csv or .json)", ext)
	}
}

// parseCsvFile parses a .csv file with headers into HostEntry objects.
// Expected headers: host, hostname, user, port, context, password
// - host: The SSH Host identifier (required, what users type to connect)
// - hostname: The SSH HostName (optional, resolved address - use for IP when DNS unavailable)
func parseCsvFile(path string) ([]HostEntry, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer func() { _ = f.Close() }()

	reader := csv.NewReader(f)

	// Read header
	header, err := reader.Read()
	if err != nil {
		if err == io.EOF {
			return nil, fmt.Errorf("CSV file is empty")
		}
		return nil, fmt.Errorf("read CSV header: %w", err)
	}

	// Build column index map
	colIdx := make(map[string]int)
	for i, col := range header {
		colIdx[strings.ToLower(strings.TrimSpace(col))] = i
	}

	// Host is required
	if _, ok := colIdx["host"]; !ok {
		return nil, fmt.Errorf("CSV file missing required 'host' column")
	}

	var entries []HostEntry
	lineNum := 1 // header was line 1

	for {
		lineNum++
		record, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("line %d: %w", lineNum, err)
		}

		entry := HostEntry{}

		// Get host (required) - the SSH Host identifier
		if idx, ok := colIdx["host"]; ok && idx < len(record) {
			entry.Host = strings.TrimSpace(record[idx])
		}
		if entry.Host == "" {
			continue // Skip empty rows
		}

		// Get optional hostname - the SSH HostName (resolved address)
		if idx, ok := colIdx["hostname"]; ok && idx < len(record) {
			entry.HostName = strings.TrimSpace(record[idx])
		}
		if idx, ok := colIdx["user"]; ok && idx < len(record) {
			entry.User = strings.TrimSpace(record[idx])
		}
		if idx, ok := colIdx["port"]; ok && idx < len(record) {
			if portStr := strings.TrimSpace(record[idx]); portStr != "" {
				port, err := strconv.Atoi(portStr)
				if err != nil {
					return nil, fmt.Errorf("line %d: invalid port '%s'", lineNum, portStr)
				}
				entry.Port = port
			}
		}
		if idx, ok := colIdx["context"]; ok && idx < len(record) {
			entry.Context = strings.TrimSpace(record[idx])
		}
		if idx, ok := colIdx["password"]; ok && idx < len(record) {
			entry.Password = strings.TrimSpace(record[idx])
		}

		entries = append(entries, entry)
	}

	return entries, nil
}

// parseJsonFile parses a .json file (array of objects) into HostEntry objects.
func parseJsonFile(path string) ([]HostEntry, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var entries []HostEntry
	if err := json.Unmarshal(data, &entries); err != nil {
		return nil, fmt.Errorf("parse JSON: %w", err)
	}

	// Filter out entries with empty host
	var filtered []HostEntry
	for _, e := range entries {
		if e.Host != "" {
			filtered = append(filtered, e)
		}
	}

	return filtered, nil
}
